### PR TITLE
python3Packages.mlx: init at 0.1.0

### PR DESCRIPTION
**Pull request by viraptor**
## Description of changes

Add the mlx library which is a new optional dependency for safetensors and will be needed for https://github.com/NixOS/nixpkgs/pull/268378 to work on darwin.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/mlx/default.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+, cmake
+, xcbuild
+, zsh
+, darwin
+, blas
+, lapack
+}:
+
+let
+  # static dependencies included directly during complication
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/mlx/default.nix:16]:
```suggestion
  # static dependencies included directly during compilation
```

Except if it was on purpose...

File: pkgs/development/python-modules/mlx/default.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  gguflib = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/mlx/default.nix:17]: why not

```suggestion
  gguf-tools = fetchFromGitHub {
```

?

File: pkgs/development/python-modules/mlx/default.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    owner = "antirez";
+    repo = "gguf-tools";
+    rev = "af7d88d808a7608a33723fba067036202910acb3";
+    hash = "sha256-LqNvnUbmq0iziD9VP5OTJCSIy+y/hp5lKCUV7RtKTvM=";
+  };
+  json = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/mlx/default.nix:23]: also, a better name:

```suggestion
  nlohmann_json = fetchFromGitHub {
```

File: pkgs/development/python-modules/mlx/default.nix [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    PYPI_RELEASE = version;
+    # we can't use Metal compilation with Darwin SDK 11
+    CMAKE_ARGS = "-DMLX_BUILD_METAL=OFF -DFETCHCONTENT_FULLY_DISCONNECTED=TRUE -DFETCHCONTENT_SOURCE_DIR_GGUFLIB=${gguflib} -DFETCHCONTENT_SOURCE_DIR_JSON=${json}";
+  };
+
+  nativeBuildInputs = with python3Packages; [ setuptools cmake pybind11 xcbuild zsh gguflib json ];
+
+  buildInputs = [ blas lapack ];
+
+  src = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/mlx/default.nix:58]: src after version

File: pkgs/development/python-modules/mlx/default.nix [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  dontUseCmakeConfigure = true;
+
+  env = {
+    PYPI_RELEASE = version;
+    # we can't use Metal compilation with Darwin SDK 11
+    CMAKE_ARGS = "-DMLX_BUILD_METAL=OFF -DFETCHCONTENT_FULLY_DISCONNECTED=TRUE -DFETCHCONTENT_SOURCE_DIR_GGUFLIB=${gguflib} -DFETCHCONTENT_SOURCE_DIR_JSON=${json}";
+  };
+
+  nativeBuildInputs = with python3Packages; [ setuptools cmake pybind11 xcbuild zsh gguflib json ];
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/mlx/default.nix:54]:
```suggestion
  nativeBuildInputs = [ cmake xcbuild zsh gguflib json ] ++ (with python3Packages; [ setuptools pybind11 ]);
```

File: pkgs/development/python-modules/mlx/default.nix [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    substituteInPlace CMakeLists.txt \
+      --replace "/usr/bin/xcrun" "${xcbuild}/bin/xcrun" \
+  '';
+
+  dontUseCmakeConfigure = true;
+
+  env = {
+    PYPI_RELEASE = version;
+    # we can't use Metal compilation with Darwin SDK 11
+    CMAKE_ARGS = "-DMLX_BUILD_METAL=OFF -DFETCHCONTENT_FULLY_DISCONNECTED=TRUE -DFETCHCONTENT_SOURCE_DIR_GGUFLIB=${gguflib} -DFETCHCONTENT_SOURCE_DIR_JSON=${json}";
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/mlx/default.nix:51]:
```suggestion
    CMAKE_ARGS = toString [ 
      (lib.cmakeBool "MLX_BUILD_METAL" false)
      (lib.cmakeBool "FETCHCONTENT_FULLY_DISCONNECTED" true)
      (lib.cmakeOptionType "filepath" "FETCHCONTENT_SOURCE_DIR_GGUFLIB" gguflib)
      (lib.cmakeOptionType "filepath" "FETCHCONTENT_SOURCE_DIR_JSON" json)
    ];
```

P.S.: should those be put on nativeBuildInputs too?

File: pkgs/development/python-modules/mlx/default.nix [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    rev = "refs/tags/v${version}";
+    hash = "sha256-xNJPG8XGbC0fy6RGcn1cxCsejyHsgnV35PuP8F1I4R4=";
+  };
+
+  meta = with lib; {
+    homepage = "https://github.com/ml-explore/mlx";
+    description = "An array framework for Apple silicon";
+    changelog = "https://github.com/ml-explore/mlx/releases/tag/v${version}";
+    license = licenses.mit;
+    platforms = [ "aarch64-darwin" ];
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/mlx/default.nix:70]: Doesn't work on new Apple?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/mlx/default.nix (hunk 7 above)
Comment (viraptor) [reply pkgs/development/python-modules/mlx/default.nix:70]: It works only on new apple.

**Pull request merged**